File: python/test/t_NormalityTest_std.py
#! /usr/bin/env python

from openturns import *

TESTPREAMBLE()
RandomGenerator().SetSeed(0)

try :

    continuousDistributionCollection = DistributionCollection()
    discreteDistributionCollection = DistributionCollection()
    distributionCollection = DistributionCollection()

    beta = Beta(2.,3.,0.,1.)
    distributionCollection.add(beta)
    continuousDistributionCollection.add(beta)

    gamma = Gamma(1.,2.,3.)
    distributionCollection.add(gamma)
    continuousDistributionCollection.add(gamma)

    gumbel = Gumbel(1.,2.)
    distributionCollection.add(gumbel)
    continuousDistributionCollection.add(gumbel)

    lognormal = LogNormal(1.,1.,2.)
    distributionCollection.add(lognormal)
    continuousDistributionCollection.add(lognormal)

    logistic = Logistic(1.,1.)
    distributionCollection.add(logistic)
    continuousDistributionCollection.add(logistic)

    normal = Normal(1.,2.)
    distributionCollection.add(normal)
    continuousDistributionCollection.add(normal)

    truncatednormal = TruncatedNormal(1.,1.,0.,3.)
    distributionCollection.add(truncatednormal)
    continuousDistributionCollection.add(truncatednormal)

    student = Student(10.,10.)
    distributionCollection.add(student)
    continuousDistributionCollection.add(student)

    triangular = Triangular(-1.,2.,4.)
    distributionCollection.add(triangular)
    continuousDistributionCollection.add(triangular)

    uniform = Uniform(1.,2.)
    distributionCollection.add(uniform)
    continuousDistributionCollection.add(uniform)

    weibull = Weibull(1., 1., 2.)
    distributionCollection.add(weibull)
    continuousDistributionCollection.add(weibull)

    geometric = Geometric(.5)
    distributionCollection.add(geometric)
    discreteDistributionCollection.add(geometric)

    poisson = Poisson(2.)
    distributionCollection.add(poisson)
    discreteDistributionCollection.add(poisson)

    collection = UserDefinedPairCollection(3, UserDefinedPair(NumericalPoint(1), 0.0))

    point = NumericalPoint(1)
    point[0] = 1.0
    collection[0] = UserDefinedPair(point, 0.3)
    point[0] = 2.0
    collection[1] = UserDefinedPair(point, 0.2)
    point[0] = 3.0
    collection[2] = UserDefinedPair(point, 0.5)
    userdefined = UserDefined(collection)
    distributionCollection.add(userdefined)
    discreteDistributionCollection.add(userdefined)

    size = 10000

    # Number of continuous distributions
    continuousDistributionNumber = continuousDistributionCollection.getSize()
    # Number of discrete distributions
    discreteDistributionNumber = discreteDistributionCollection.getSize()
    # Number of distributions
    distributionNumber = continuousDistributionNumber + discreteDistributionNumber

    # We create a collection of NumericalSample of size "size" and of dimension 1 (scalar values) : the collection has distributionNumber NumericalSamples
    sampleCollection = [NumericalSample(size, 1) for i in range(distributionNumber)]
    # We create a collection of NumericalSample of size "size" and of dimension 1 (scalar values) : the collection has continuousDistributionNumber NumericalSamples
    continuousSampleCollection = [NumericalSample(size, 1) for i in range(continuousDistributionNumber)]
    # We create a collection of NumericalSample of size "size" and of dimension 1 (scalar values) : the collection has discreteDistributionNumber NumericalSamples
    discreteSampleCollection = [NumericalSample(size, 1) for i in range(discreteDistributionNumber)]

    for i in range(continuousDistributionNumber) :
        continuousSampleCollection[i] = continuousDistributionCollection[i].getSample(size)
        continuousSampleCollection[i].setName(continuousDistributionCollection[i].getName())
        sampleCollection[i] = continuousSampleCollection[i]
    for i in range(discreteDistributionNumber) :
        discreteSampleCollection[i] = discreteDistributionCollection[i].getSample(size)
        discreteSampleCollection[i].setName(discreteDistributionCollection[i].getName())
        sampleCollection[continuousDistributionNumber + i] = discreteSampleCollection[i]

    # Test the normality of several samples using the Anderson Darling test
    andersonDarlingResult = NumericalPoint(distributionNumber)
    for i in range(distributionNumber) :
        result = NormalityTest().AndersonDarlingNormal(sampleCollection[i])
        andersonDarlingResult[i] = result.getBinaryQualityMeasure()
        print "sample " , sampleCollection[i].getName() , " result=" , andersonDarlingResult[i]
    print "andersonDarlingResult=" , repr(andersonDarlingResult)

    # Test the normality of several samples using the Cramer Von Mises test
    cramerVonMisesResult = NumericalPoint(distributionNumber)
    for i in range(distributionNumber) :
        result = NormalityTest().CramerVonMisesNormal(sampleCollection[i])
        cramerVonMisesResult[i] = result.getBinaryQualityMeasure()
        print "sample " , sampleCollection[i].getName() , " result=" , cramerVonMisesResult[i]
    print "cramerVonMisesResult=" , repr(cramerVonMisesResult)

except :
    import sys
    print "t_NormalityTest_std.py", sys.exc_type, sys.exc_value

File: lib/test/t_NormalityTest_std.cxx
//                                               -*- C++ -*-
/**
 *  @file  t_NormalityTest_std.cxx
 *  @brief The test file of class Test
 *
 *  Copyright (C) 2005-2013 EDF-EADS-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 *  @author schueller
 *  @date   2012-07-16 15:59:45 +0200 (Mon, 16 Jul 2012)
 */
#include "OT.hxx"
#include "OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

int main(int argc, char *argv[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);
  setRandomGenerator();

  Collection<Distribution> continuousDistributionCollection;
  Collection<Distribution> discreteDistributionCollection;
  Collection<Distribution> distributionCollection;

  Beta beta(2., 3., 0., 1.);
  distributionCollection.add(beta);
  continuousDistributionCollection.add(beta);

  Gamma gamma(1., 2., 3.);
  distributionCollection.add(gamma);
  continuousDistributionCollection.add(gamma);

  Gumbel gumbel(1., 2.);
  distributionCollection.add(gumbel);
  continuousDistributionCollection.add(gumbel);

  LogNormal lognormal(1., 1., 2.);
  distributionCollection.add(lognormal);
  continuousDistributionCollection.add(lognormal);

  Logistic logistic(1., 1.);
  distributionCollection.add(logistic);
  continuousDistributionCollection.add(logistic);

  Normal normal(1., 2.);
  distributionCollection.add(normal);
  continuousDistributionCollection.add(normal);

  TruncatedNormal truncatednormal(1., 1., 0., 3.);
  distributionCollection.add(truncatednormal);
  continuousDistributionCollection.add(truncatednormal);

  Student student(4., 10.);
  distributionCollection.add(student);
  continuousDistributionCollection.add(student);

  Triangular triangular(-1., 2., 4.);
  distributionCollection.add(triangular);
  continuousDistributionCollection.add(triangular);

  Uniform uniform(1., 2.);
  distributionCollection.add(uniform);
  continuousDistributionCollection.add(uniform);

  Weibull weibull(1., 1., 2.);
  distributionCollection.add(weibull);
  continuousDistributionCollection.add(weibull);

  Geometric geometric(.5);
  distributionCollection.add(geometric);
  discreteDistributionCollection.add(geometric);

  Poisson poisson(2.);
  distributionCollection.add(poisson);
  discreteDistributionCollection.add(poisson);

  Collection<UserDefinedPair> collection(3, UserDefinedPair(NumericalPoint(1), 0.0));

  NumericalPoint point(1);
  point[0] = 1.0;
  collection[0] = UserDefinedPair(point, 0.3);
  point[0] = 2.0;
  collection[1] = UserDefinedPair(point, 0.2);
  point[0] = 3.0;
  collection[2] = UserDefinedPair(point, 0.5);
  UserDefined userdefined(collection);
  distributionCollection.add(userdefined);
  discreteDistributionCollection.add(userdefined);

  UnsignedLong size(10000);

  UnsignedLong continuousDistributionNumber(continuousDistributionCollection.getSize());
  UnsignedLong discreteDistributionNumber(discreteDistributionCollection.getSize());
  UnsignedLong distributionNumber(continuousDistributionNumber + discreteDistributionNumber);
  Collection<NumericalSample> sampleCollection(distributionNumber);
  Collection<NumericalSample> continuousSampleCollection(continuousDistributionNumber);
  Collection<NumericalSample> discreteSampleCollection(discreteDistributionNumber);
  for (UnsignedLong i = 0; i < continuousDistributionNumber; i++)
    {
      continuousSampleCollection[i] = continuousDistributionCollection[i].getSample(size);
      continuousSampleCollection[i].setName(continuousDistributionCollection[i].getName());
      sampleCollection[i] = continuousSampleCollection[i];
    }
  for (UnsignedLong i = 0; i < discreteDistributionNumber; i++)
    {
      discreteSampleCollection[i] = discreteDistributionCollection[i].getSample(size);
      discreteSampleCollection[i].setName(discreteDistributionCollection[i].getName());
      sampleCollection[continuousDistributionNumber + i] = discreteSampleCollection[i];
    }
  // Test the normality of several samples using the Anderson Darling test
  NumericalPoint andersonDarlingResult(distributionNumber);
  for (UnsignedLong i = 0; i < distributionNumber; i++)
    {
      TestResult result(NormalityTest::AndersonDarlingNormal(sampleCollection[i]));
      andersonDarlingResult[i] = result.getBinaryQualityMeasure();
      fullprint << "sample " << sampleCollection[i].getName() << " result=" << andersonDarlingResult[i] << std::endl;
    }
  fullprint << "andersonDarlingResult=" << andersonDarlingResult << std::endl;
  // Test the normality of several samples using the Cramer Von Mises test
  NumericalPoint cramerVonMisesResult(distributionNumber);
  for (UnsignedLong i = 0; i < distributionNumber; i++)
    {
      TestResult result(NormalityTest::CramerVonMisesNormal(sampleCollection[i]));
      cramerVonMisesResult[i] = result.getBinaryQualityMeasure();
      fullprint << "sample " << sampleCollection[i].getName() << " result=" << cramerVonMisesResult[i] << std::endl;
    }
  fullprint << "cramerVonMisesResult=" << cramerVonMisesResult << std::endl;
  return ExitCode::Success;
}
